pkg/hiveregistrator/registrator.go

```diff
@@ -0,0 +1,100 @@
+package hiveregistrator
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/gofrs/uuid"
+	hivev1 "github.com/openshift/hive/apis/hive/v1"
+	hiveclient "github.com/openshift/hive/pkg/client/clientset/versioned"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/util/retry"
+)
+
+const HIVEENVVARIABLE = "HIVEKUBECONFIGPATH"
+
+type Registrator interface {
+	Register(ctx context.Context, workloadCluster *WorkloadCluster) (*hivev1.ClusterDeployment, error)
+	IsConnected(ctx context.Context, namespaceName string) (bool, string, error)
+}
+
+// Represents all data in hive pertaining to a single ARO cluster
+type WorkloadCluster struct {
+	SubscriptionID              string `json:"subscription,omitempty"`
+	ClusterName                 string `json:"name,omitempty"`
+	ResourceGroupName           string `json:"resourceGroup,omitempty"`
+	Location                    string `json:"location,omitempty"`
+	BaseDomain                  string `json:"basedomain,omitempty"`
+	BaseDomainResourceGroupName string `json:"basedomainResourceGroupName,omitempty"`
+	InfraID                     string `json:"infraId,omitempty"`
+	ClusterID                   string `json:"clusterID,omitempty"`
+	KubeConfig                  string `json:"kubeconfig,omitempty"`
+	ServicePrincipal            string `json:"serviceprincipal,omitempty"`
+}
+
+type hiveRegistrator struct {
+	hiveClientset *hiveclient.Clientset
+	kubernetescli *kubernetes.Clientset
+}
+
+func New(hiveClientset *hiveclient.Clientset, kubernetescli *kubernetes.Clientset) Registrator {
+	return &hiveRegistrator{
+		hiveClientset: hiveClientset,
+		kubernetescli: kubernetescli,
+	}
+}
+
+func (hr *hiveRegistrator) Register(ctx context.Context, workloadCluster *WorkloadCluster) (*hivev1.ClusterDeployment, error) {
+	var namespaceName string
+
+	err := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		namespaceName = "aro-" + uuid.Must(uuid.NewV4()).String()
+		csn := ClusterNamespace(namespaceName)
+		_, err := hr.kubernetescli.CoreV1().Namespaces().Create(ctx, csn, metav1.CreateOptions{})
+		return err
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	kubesecret := KubeAdminSecret(namespaceName, []byte(workloadCluster.KubeConfig))
+
+	_, err = hr.kubernetescli.CoreV1().Secrets(namespaceName).Create(ctx, kubesecret, metav1.CreateOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	spsecret := ServicePrincipalSecret(namespaceName, []byte(workloadCluster.ServicePrincipal))
```

Comment:
we need to patch the `Update` workflow to update this secret as well.  see `pkg/cluster/install.go -> Update` function

Reply:
I created a work item for this: https://msazure.visualstudio.com/AzureRedHatOpenShift/_workitems/edit/14895480

Let's do it separately from this PR.

Good catch @bennerv 👍